Azure.MgmtSdk.Analyzers/ModelNameSuffixCondition.cs

```diff
@@ -0,0 +1,168 @@
+        private static readonly string Description = "Suffix is not recommended. Consider to remove or modify.";
+
+        private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title,
+            MessageFormat, DiagnosticCategory.Naming, DiagnosticSeverity.Warning, isEnabledByDefault: true,
+            description: Description);
+
+        // ConditionOne: Avoid using Definition as model suffix unless it's the name of a Resource
+        private static readonly Regex SuffixRegexConditionOne = new Regex(".+(?<Suffix>(Definition?))$");
+        // ConditionTwo: Avoid using Data as model suffix unless the model derives from ResourceData/TrackedResourceData
+        private static readonly Regex SuffixRegexConditionTwo = new Regex(".+(?<Suffix>(Data))$");
+        // ConditionThree: Avoid using Operation as model suffix unless the model derives from Operation<T>
+        private static readonly Regex SuffixRegexConditionThree = new Regex(".+(?<Suffix>(Operation))$");
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.EnableConcurrentExecution();
+            context.RegisterSymbolAction(AnalyzeSuffixConditionOne, SymbolKind.NamedType);
+            context.RegisterSymbolAction(AnalyzeSuffixConditionTwo, SymbolKind.NamedType);
+            context.RegisterSymbolAction(AnalyzeSuffixConditionThree, SymbolKind.NamedType);
+        }
+
+
+        public static bool ImplementsInterfaceOrBaseClass(INamedTypeSymbol typeSymbol, String nameToCheck)
+        {
+            if (typeSymbol == null)
+                return false;
+
+            if (typeSymbol.MetadataName == nameToCheck)
+                return true;
+
+            if (typeSymbol.BaseType.MetadataName == nameToCheck)
+                return true;
+
+            foreach (var @interface in typeSymbol.AllInterfaces)
+                if (@interface.MetadataName == nameToCheck)
+                    return true;
+
+            return false;
+        }
+
+        private void AnalyzeSuffixConditionOne(SymbolAnalysisContext context)
+        {
+            var name = context.Symbol.Name;
+            var typeSymbol = (INamedTypeSymbol)context.Symbol;
+            if (typeSymbol.TypeKind != TypeKind.Class) // We just check classes.
+                return;
+
+            var match = SuffixRegexConditionOne.Match(name);
+            if (match.Success)
+            {
+                bool hasNamespaceModels = false;
+                for (var namespaceSymbol = typeSymbol.ContainingNamespace; namespaceSymbol != null; namespaceSymbol = namespaceSymbol.ContainingNamespace)
```

Comment:
This is a repeated logic, consider extract into a utility method. I remember it's used in `ModelNameSuffix` as well. So better to extract into a parent class (if any), or into a static method.